spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/condition/OnClassCondition.java

```diff
@@ -138,10 +140,20 @@ private static final class ThreadedOutcomesResolver implements OutcomesResolver
 
 		private final Thread thread;
 
+		private final Lock outcomesLock = new ReentrantLock();
+
 		private volatile ConditionOutcome[] outcomes;
```

Comment:
`volatile` on arrays guarantees that the reference to the array is refreshed but it does not guarantee that the array contents are visible to another thread. I used a `Lock` to get a read/write barriers to make sure that the array contents are visible.